### DISCUSSION TITLE
💥Breaking change in v0.22.1

### DISCUSSION THREAD

**Discussion by earthboundkid** [earthboundkid/requests#31] [Announcements]
v0.22.1 [removes .Get and .Post](https://github.com/carlmjohnson/requests/commit/4ce2ff9d57ff443bb9d32226f9767f417938b92e). In almost all cases, calls to these methods should be redundant and can be simply dropped. In the cases where it was not, use `Method(http.MethodGet)` or `Method(http.MethodPost)` instead.

**Comment by stephenafamo** [comment]
Why not also then drop other similar methods: `Put()`, `Patch()`, `Head()`, `Delete()`?

**Reply by earthboundkid** [reply]
GET and POST are automatically set by the presence or absence of a body, so you don’t need a convenience method for them. The other methods have to be set by hand, so it’s nice to have an easy way to do it.

**Reply by stephenafamo** [reply]
Since the `Method()` function exists, I think the rest should be deprecated/removed.

In general, I feel like the Go way is to have one way of doing something.

Also, using the `Method()` function is just as simple IMO.

**Reply by earthboundkid** [reply]
Everything in [`builder_extras.go`](https://github.com/carlmjohnson/requests/blob/main/builder_extras.go) is a redundant second way to do things. It’s just convenient to have them. If this was a part of the Go standard library, I would agree, but for a third party package, it’s nice to have a little more hand holding.

**Reply by stephenafamo** [reply]
Alright. Makes sense 👍🏾

**Reply by earthboundkid** [reply]
I would like to have a better way of separating the “core” features from the convenient extras, but so far I’ve only been able to just move them to separate files. I tried to make them different types but that screwed up the method chaining. If you think of a better way to make it clear what it core and what’s extra, let me know.

**Reply by stephenafamo** [reply]
My suggestion would need a whole redesign.

That would be to use option arguments that satisfy the interface:

```go
type BuilderOption interface {
    Apply(b *Builder)
}
```

Then instead of method chaining, it would have a constructor like: 

```go
func New(baseURL string, opts ...BuilderOption) *Builder {
    // ...
}
```

And then `Builder` will only have the other methods like `Request()`, `Fetch()` and `Do()`.

This then allows for the following:

1. You can move all the helper options into a separate package. Or even create a whole new module for it.
2. Users of the package can create their own helper options to use with it.

**Reply by stephenafamo** [reply]
The developer experience of this method is quite similar.  
Using this example from the README.

```go
err := requests.
	URL("https://postman-echo.com/post").
	BodyBytes([]byte(`hello, world`)).
	ContentType("text/plain").
	Fetch(context.Background())
```

That becomes

```go
err := requests.New(
            "https://postman-echo.com/post",
	    requests.BodyBytes([]byte(`hello, world`)),
	    requests.ContentType("text/plain"),
	).Fetch(context.Background())
```

If someone really wanted to make it look the same, they could use a dot import but I don't think that is necessary.

**Reply by earthboundkid** [reply]
Note that the Config function already does basically that. I should make it take a variadic slice of config functions instead of just one though. I’ve written about it elsewhere, so I won’t get into it here, but for this project I think the functional option pattern isn’t as good as method chaining.

**Reply by stephenafamo** [reply]
That makes sense. 

Along with making it take a variadic slice, I think it would be better if Config was an interface rather than a raw function. Makes it generally easier to use IMO.

You can also add a `ConfigFunc` type which can satisfy the interface, similar to `http.Handler` and `http.HandlerFunc`
